### PR TITLE
[Bug-fix][1.6] Improve QAT accuracy

### PR DESCRIPTION
**Pull request by Sand3r-**
Provides a fix for accuracy drop obtained with QAT models by enabling uint8 variable type on conv-relu outputs.
Also enables executing VGG16 and VGG19 by:
* Adding additional way of fetching scales (by associating tensor with its tensor scale by the *.scale suffix)
* Fixing bug with empty list of disabled op ids
Restores fake-quant mul to achieve best accuracy.

### Resnet50
|                 | top1 acc | top 5 acc |
|-----------------|----------|-----------|
| Original QAT    | 0.7655   | 0.9304    |
| Transformed QAT | 0.7653   | 0.9298    |
| Diff            | 0.0002   | 0.0006    |

### Mobilenet V1
|                 | top1 acc | top 5 acc |
|-----------------|----------|-----------|
| Original QAT    | 0.7077   | 0.8954    |
| Transformed QAT | 0.7076   | 0.8943    |
| Diff            | 0.0001   | 0.0011    |

### CODE REVIEW

**Comment by Sand3r-** [comment]
For some reason the CI fails on `test_parallel_executor_seresnext_with_reduce_gpu` due to accuracy diff. I'm not sure if there is anything related to it, what has been proposed in this PR.

**Comment by luotao1** [comment]
Please create PR to develop first, and after merge it to develop, then cherry-pick to release/1.6.

**Comment by luotao1** [comment]
You can ignore the `test_parallel_executor_seresnext_with_reduce_gpu` random fail.

**Comment by Sand3r-** [comment]
@luotao1 Thank you for the information. I have rebased on top of develop, and changed the target branch in this PR to it.

**Comment by luotao1** [comment]
Please update the code format to pass the travis_ci

**Comment by wzzju** [comment]
> Provides a fix for accuracy drop obtained with QAT models by enabling uint8 variable type on conv-relu outputs.
> Also enables executing VGG16 and VGG19 by:
> 
> * Adding additional way of fetching scales (by associating tensor with its tensor scale by the *.scale suffix)
> * Fixing bug with empty list of disabled op ids
>   Restores fake-quant mul to achieve best accuracy.
> 
> ## Accuracy comparison
> ### Resnet50
> top1 acc	top 5 acc
> Original QAT	0.7655	0.9304
> Transformed QAT	0.7653	0.9298
> Diff	0.0002	0.0006
> ### Mobilenet V1
> top1 acc	top 5 acc
> Original QAT	0.7077	0.8954
> Transformed QAT	0.7076	0.8943
> Diff	0.0001	0.0011

1. What is the original bug? Could you please give more information about this bug?
2. How do you resolve this bug? Could you please give more details?
3. Could you please explain what is `enabling uint8 variable type on conv-relu outputs` ? Because QAT doesn't provide the uint8 scale, how do you enable uint8 variable type on conv-relu outputs?

**Comment by Sand3r-** [comment]
@wzzju 
> What is the original bug? Could you please give more information about this bug?

Certainly. Based on parameters such as `fuse_residual_connection` and `fuse_activation` int8 convolution kernel makes a decision what should the output datatype be. Since in many cases 'relu' activation was fused with convolution, the output was set to be uint8, however the signed scale has been assigned to it.

> How do you resolve this bug? Could you please give more details?

To resolve this bug, I have changed the information about the scale in `self._var_quant_scales` map that the scale provided to the `cpu_quantize_pass` should be rescaled to work with uint8 data. This is accomplished by changing the first value of the pair (bool unsigned, Tensor scale) to true, as done in line https://github.com/PaddlePaddle/Paddle/pull/20174/files#diff-eba5e4d0f5420f0702b898665a44dc56R571

To shed more light on  how this done: `cpu_quantize_pass` accepts the `VarQuantScales` map where the scales provided are "normalized". Based on the `unsigned` parameter they are then either multiplied by max_s8 (127) or max_u8 (255) so that later on they can be used in int8 computation kernels without further modifications.


> Could you please explain what is enabling uint8 variable type on conv-relu outputs ? Because QAT doesn't provide the uint8 scale, how do you enable uint8 variable type on conv-relu outputs?
As described above, the conv (with fused relu) output scales provided by QAT are appropriately rescaled.

File: python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py [hunk 1]
```diff
@@ -557,13 +560,16 @@ def _transform_to_quantize_mkldnn(self, graph, op_node):
...
-        op_node.set_attr("force_fp32_output", True)
+    def _update_conv_relu_scales(self, graph):
+        for op in graph.all_op_nodes():
+            if op.name() in self._conv_ops:
+                out_name = op.output("Output")[0]
+                if out_name in self._var_quant_scales and \
+                   op.op().attr("fuse_activation") == 'relu' and \
+                   op.op().attr("fuse_residual_connection") == False:
+                    _, tensor = self._var_quant_scales[out_name]
+                    self._var_quant_scales[out_name] = (True, tensor)
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py:571]: Here, you just set the first flag as `True`, but don't change the scale value. Is it right?

Comment (Sand3r-) [reply python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py:571]: Yes, here I pass the normalized scale. The actual rescaling happens here:
https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/framework/ir/mkldnn/cpu_quantize_pass.cc#L46-L47

`scale_to_one` is what we've passed to the quantizer.

File: python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py [hunk 2]
```diff
@@ -350,6 +351,8 @@ def _gather_scales(self, graph):
                 use_unsigned_int = False
                 self._var_quant_scales[input_name] = (use_unsigned_int,
                                                       lod_tensor)
+                self._var_quant_scales[scale_name.replace(".scale", "")] = (
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py:354]: What is this prepare for? Are `input_name` and `scale_name.replace(".scale", "")` different？

Comment (Sand3r-) [reply python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py:354]: I wish they were the same, but when we encounter the follwing case:
[op]->[fake_dequantize]->[fake_quantize]->[another_op] 
as it is in, for example, VGG16, the tensor that enters `fake_quantize_moving_average_abs_max` has a `.dequantized` suffix.
image
In this example `pool2d_1.tmp_0.dequantized` enters `fake_quantize_moving_average_abs_max`.

Comment (Sand3r-) [reply python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py:354]: The reason for which we need the clear names without any *.dequantized, or *.scale suffixes, is that after the graph is cleared off of all the fake-(de)quantize operators, the tensors with suffixes disappear, and `cpu_quantize_pass` cannot find appropriate scales for tensors with normal names, such as `pool2d_1.tmp_0`.

Comment (wzzju) [reply python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py:354]: A very clear explanation. :thumbsup:

**Review by wzzju — APPROVED** [review]
LGTM.

**Pull request merged**